av.h

```diff
@@ -112,64 +112,96 @@ If all you need is to look up an array element, then prefer C<av_fetch>.
+
+=over
+
+=item C<newAV> form
+
+=for comment
+'form' above and below is because otherwise have two =items with the same name,
+can't link to them.
+
+This does nothing beyond creating the whole-array data structure.
+The Perl equivalent is approximately S<C<my @array;>>
 
-This is very useful when an AV is required, but populating it may be
-deferred, or even never actually take place. (Memory is not allocated
-unnecessarily.)
+This is useful when the minimum size of the array could be zero (perhaps there
+are likely code paths that will entirely skip using it).
 
-Subsequent SV* array allocation would be performed via C<L</av_extend>>.
-This might be called directly:
-    av_extend(av, key);
+If the array does get used, the pointers data structure will need to be
+allocated at that time.  This will end up being done by L</av_extend>>,
+either explicitly:
+
+    av_extend(av, len);
+
+or implicitly when the first element is stored:
 
-Or it might be called implicitly when the first element is stored:
     (void)av_store(av, 0, sv);
 
 Unused array elements are typically initialized by C<av_extend>. (Only
 core maintainers should have need to concern themseleves with when that
-is not the case. Refer to F<av.h> and F<av.c> for the differences between
+is not the case.
+
+Refer to F<av.h> and F<av.c> for the differences between
 real and fake AVs.)
 
-In contrast, when an AV is created for immediate population with a known
-(or likely) number of elements, it is more efficient to immediately
-allocate a SV* array of the necessary size. (This avoids inefficient use
-of C<av_extend> and the potential for the first allocation being too small
-and then having to resize it.)
+=item C<newAV_alloc_x> form
+
+This effectively does a C<newAV> followed by also allocating (uninitialized)
+space for the pointers array.  This is used when you know ahead of time the
+likely minimum size of the array.  It is more efficient to do this than doing a
+plain C<newAV> followed by an C<av_extend>.
+
+Of course the array can be extended later should it become necessary.
+
+C<size> must be at least 1.
+
+=item C<newAV_alloc_xz> form
 
-For that scenario, newAV_alloc_x and newAV_alloc_xz can be used to create
-an AV and allocate a SV* array to fit the specified number of elements.
-(As a result, these macros MUST NOT be called with a size less than 1.)
+This is C<newAV_alloc_x>, but initializes each pointer in it to NULL.  This
+gives added safety to guard against them being read before being set.  If you
+know now what those elements should be, instead use L</C<av_make>>.
 
-newAV_alloc_x does not initialize the array elements - and so the
-expectation is that all will be initialized elsewhere prior to any
-potentials reads. newAV_alloc_xz does initialize the array elements.
+C<size> must be at least 1.
+
+=back
 
 The following examples all result in an array that can fit four elements
 (indexes 0 .. 3):
 
-    AV *av = newAV();
-    av_extend(av, 1);
-
```

Comment:
This example was intended to point out, albeit opaquely, that av_extend() forces a minimum pointer array size. I don't hugely object to this change though.

Reply:
It was too opaque in my opinion, and led to questions about how much do you get by default, and sort of committed us to never lowering the default below 4.  That's why I omitted it.  Language explicitly stating there is a default minimum might be helpful.  I don't know